packages/hardhat/test/QuestFactory.test.ts

```diff
@@ -1,47 +1,173 @@
+    otherToken = await tokenMockFactory.deploy("Other Token", "OTOKEN");
+    questFactoryContract = await new QuestFactory__factory(owner).deploy(
+      owner.address,
+      depositToken.address,
+      depositAmount
+    );
+    await depositToken.connect(owner).mint(owner.address, fromNumber(1000));
+  });
+
+  it("SHOULD set the owner address correctly", async function () {
     expect(!!owner.address).to.eq(true); // truthy
   });
 
   describe("createQuest()", function () {
-    let questFactoryContract;
-
-    beforeEach(async function () {
-      await deployments.fixture(["TokenMock", "QuestFactory"]);
-      questFactoryContract = await ethers.getContract("QuestFactory");
-    });
-
-    it("should emit QuestCreated", async function () {
+    it("SHOULD emit QuestCreated", async function () {
       // Arrange
-      const tokenContract = await deployments.get("TokenMock");
       const title = "title";
       const detailIPFS = "0x";
       const expireTime = 0; // Unix Epoch 0
+      await depositToken
+        .connect(owner)
+        .approve(questFactoryContract.address, depositAmount);
 
       // Act
       // Assert
       expect(
         await questFactoryContract.createQuest(
           title,
           detailIPFS,
-          tokenContract.address,
+          rewardToken.address,
           expireTime,
           owner.address
         )
       ).to.emit(questFactoryContract, "QuestCreated");
     });
+
+    it("SHOULD collect deposit and transfer to new quests", async () => {
+      // Arrange
+      const title = "title";
+      const detailIPFS = "0x";
+      const expireTime = 0; // Unix Epoch 0
+
+      await depositToken
+        .connect(owner)
+        .approve(questFactoryContract.address, depositAmount);
+
+      // Act
+      const newQuestAddress = await extractQuestAddressFromTransaction(
+        await questFactoryContract.createQuest(
+          title,
+          detailIPFS,
+          rewardToken.address,
+          expireTime,
+          owner.address
+        )
+      );
+
+      // Assert
+      expect(await depositToken.balanceOf(newQuestAddress)).to.eq(
+        depositAmount
+      );
+    });
+
+    it("SHOULD revert WHEN creator didn't aproved enough funds", async () => {
+      // Arrange
+      const title = "title";
+      const detailIPFS = "0x";
+      const expireTime = 0; // Unix Epoch 0
+
+      await depositToken
+        .connect(owner)
+        .approve(questFactoryContract.address, depositAmount.div(2));
+      // Act
+      const act = () =>
+        questFactoryContract.createQuest(
+          title,
+          detailIPFS,
+          rewardToken.address,
+          expireTime,
+          owner.address
+        );
+
+      // Assert
+      await expect(act()).to.be.revertedWith(
+        "ERC20: transfer amount exceeds allowance"
+      );
+    });
+  });
+
+  describe("setDeposit()", function () {
+    it("SHOULD emit DepositChanged WHEN valid", async () => {
```

Comment:
What happend when i setDeposit with amount = 0? What behaviour we want?

Reply:
Already implemented haha
![image](https://user-images.githubusercontent.com/13890618/166176979-4006ef2b-ce4f-444b-93db-87d70a97c173.png)

Reply:
I took these two bitches from govern contracts

Reply:
Hahah. Nice!